File: app/models/gov_meeting.py
from datetime import datetime
from typing import Optional
from sqlalchemy import BigInteger, String, Text, DateTime, SmallInteger, Integer, Date, Time, Index
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base
from app.models.base import TimestampMixin


class GovMeetingApply(Base, TimestampMixin):
    """政企约见申请主表"""
    __tablename__ = "gov_meeting_apply"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    apply_no: Mapped[str] = mapped_column(String(64), nullable=False, unique=True)
    enterprise_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    enterprise_name: Mapped[str] = mapped_column(String(200), nullable=False)
    credit_code: Mapped[str] = mapped_column(String(64), nullable=False)
    contact_name: Mapped[str] = mapped_column(String(100), nullable=False)
    contact_phone: Mapped[str] = mapped_column(String(64), nullable=False)
    topic_code: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    topic_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    meeting_level: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)  # 兼容旧字段，已拆分
    expected_level_code: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)  # 企业期望约见层级
    expected_level_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    final_level_code: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)   # 后台研判约见层级
    final_level_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    title: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)              # 约见申请标题
    meeting_content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)           # 约见内容
    discussion_item: Mapped[Optional[str]] = mapped_column(Text, nullable=True)           # 洽谈事项
    urgency_level: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)       # 紧急程度
    industry_code: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)      # 所属行业代码
    industry_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)      # 所属行业名称
    registered_address: Mapped[Optional[str]] = mapped_column(String(500), nullable=True) # 注册地址
    description: Mapped[str] = mapped_column(Text, nullable=False)  # 申请说明（兼容）
    commitment_checked: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=0)
    region_code: Mapped[str] = mapped_column(String(32), nullable=False)
    region_name: Mapped[str] = mapped_column(String(100), nullable=False)
    service_center_id: Mapped[Optional[int]] = mapped_column(BigInteger, nullable=True)
    service_center_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    status: Mapped[str] = mapped_column(String(50), nullable=False)
    reject_reason_code: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    reject_reason_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    reject_opinion: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    submitted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    accepted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    arranged_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    meeting_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)    # 实际约见时间
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    evaluated_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    finished_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    __table_args__ = (
        Index("ix_gov_meeting_apply_enterprise_id", "enterprise_id"),
        Index("ix_gov_meeting_apply_credit_code", "credit_code"),
        Index("ix_gov_meeting_apply_region_code", "region_code"),
        Index("ix_gov_meeting_apply_status", "status"),
        Index("ix_gov_meeting_apply_submitted_at", "submitted_at"),
    )


class GovMeetingAudit(Base):
    """政企约见审核/操作流水"""
    __tablename__ = "gov_meeting_audit"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    apply_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    action_type: Mapped[str] = mapped_column(String(100), nullable=False)
    action_name: Mapped[str] = mapped_column(String(100), nullable=False)
    before_status: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    after_status: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    opinion: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    operator_type: Mapped[str] = mapped_column(String(30), nullable=False)   # USER / ENTERPRISE
    operator_id: Mapped[str] = mapped_column(String(100), nullable=False)
    operator_name: Mapped[str] = mapped_column(String(100), nullable=False)
    operator_dept_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    operator_dept_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)

    __table_args__ = (
        Index("ix_gov_meeting_audit_apply_id", "apply_id"),
        Index("ix_gov_meeting_audit_action_type", "action_type"),
    )


class GovMeetingArrangement(Base, TimestampMixin):
    """政企约见安排"""
    __tablename__ = "gov_meeting_arrangement"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    apply_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    meeting_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)  # 约见日期时间
    meeting_place: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)
    meeting_method: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)   # ON_SITE/VIDEO/PHONE
    gov_contact_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    gov_contact_phone: Mapped[Optional[str]] = mapped_column(String(64), nullable=True)
    start_time: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)     # 约见开始时间
    end_time: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)       # 约见结束时间
    host_dept_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)    # 主持部门ID
    host_dept_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)  # 主持部门名称
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)                  # 备注说明
    remark: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    confirmed_flag: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=0)

    __table_args__ = (
        Index("ix_gov_meeting_arrangement_apply_id", "apply_id"),
    )


class GovMeetingParticipant(Base):
    """政企约见参与人"""
    __tablename__ = "gov_meeting_participant"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    arrangement_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    apply_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    participant_type: Mapped[str] = mapped_column(String(30), nullable=False)   # GOV / ENTERPRISE
    participant_name: Mapped[str] = mapped_column(String(100), nullable=False)
    participant_title: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    participant_dept_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)    # 所在部门ID
    participant_dept_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)  # 所在部门名称
    contact_phone: Mapped[Optional[str]] = mapped_column(String(64), nullable=True)           # 联系电话
    role_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)              # 参会角色
    sort_no: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)

    __table_args__ = (
        Index("ix_gov_meeting_participant_arrangement_id", "arrangement_id"),
        Index("ix_gov_meeting_participant_apply_id", "apply_id"),
    )


class GovMeetingRecord(Base):
    """政企约见纪要"""
    __tablename__ = "gov_meeting_record"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    apply_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    arrangement_id: Mapped[Optional[int]] = mapped_column(BigInteger, nullable=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)          # 会议内容
    conclusions: Mapped[Optional[str]] = mapped_column(Text, nullable=True)    # 会议结论
    follow_up_items: Mapped[Optional[str]] = mapped_column(Text, nullable=True) # 后续事项
    recorder_id: Mapped[str] = mapped_column(String(100), nullable=False)
    recorder_name: Mapped[str] = mapped_column(String(100), nullable=False)
    record_time: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)
    deleted_flag: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=0)

    __table_args__ = (
        Index("ix_gov_meeting_record_apply_id", "apply_id"),
        Index("ix_gov_meeting_record_arrangement_id", "arrangement_id"),
    )

File: enterprise-h5/src/api/govMeeting.ts
import { get, post } from './request'

// ── snake_case compat helpers ─────────────────────────────────────────────

function s(obj: any, ...keys: string[]): any {
  for (const k of keys) {
    if (obj[k] !== undefined && obj[k] !== null) return obj[k]
  }
  return undefined
}

// ── Interfaces ────────────────────────────────────────────────────────────

export interface GovMeetingApply {
  id: number
  applyNo: string
  apply_no?: string
  enterpriseName: string
  enterprise_name?: string
  creditCode: string
  credit_code?: string
  contactName: string
  contact_name?: string
  contactPhone: string
  contact_phone?: string
  topicCode?: string
  topic_code?: string
  topicName?: string
  topic_name?: string
  meetingLevel?: string
  meeting_level?: string
  expectedLevelCode?: string
  expected_level_code?: string
  expectedLevelName?: string
  expected_level_name?: string
  finalLevelCode?: string
  final_level_code?: string
  finalLevelName?: string
  final_level_name?: string
  title?: string
  content?: string
  meeting_content?: string
  discussionItem?: string
  discussion_item?: string
  urgencyLevel?: string
  urgency_level?: string
  industryCode?: string
  industry_code?: string
  industryName?: string
  industry_name?: string
  registeredAddress?: string
  registered_address?: string
  description?: string
  commitmentChecked?: number
  commitment_checked?: number
  regionCode: string
  region_code?: string
  regionName: string
  region_name?: string
  serviceCenterId?: number
  service_center_id?: number
  serviceCenterName?: string
  service_center_name?: string
  status: string
  rejectReasonCode?: string
  reject_reason_code?: string
  rejectReasonName?: string
  reject_reason_name?: string
  rejectOpinion?: string
  reject_opinion?: string
  submittedAt?: string
  submitted_at?: string
  acceptedAt?: string
  accepted_at?: string
  arrangedAt?: string
  arranged_at?: string
  completedAt?: string
  completed_at?: string
  evaluatedAt?: string
  evaluated_at?: string
  finishedAt?: string
  finished_at?: string
  createdAt?: string
  created_at?: string
}

export interface GovMeetingParticipant {
  id?: number
  participantType: string
  participant_type?: string
  participantName: string
  participant_name?: string
  participantTitle?: string
  participant_title?: string
  participantDeptId?: string
  participant_dept_id?: string
  participantDeptName?: string
  participant_dept_name?: string
  contactPhone?: string
  contact_phone?: string
  roleName?: string
  role_name?: string
  sortNo?: number
  sort_no?: number
}

export interface GovMeetingArrangement {
  id: number
  applyId?: number
  apply_id?: number
  meetingDate?: string
  meeting_date?: string
  startTime?: string
  start_time?: string
  endTime?: string
  end_time?: string
  meetingPlace?: string
  meeting_place?: string
  meetingMethod?: string
  meeting_method?: string
  hostDeptName?: string
  host_dept_name?: string
  govContactName?: string
  gov_contact_name?: string
  govContactPhone?: string
  gov_contact_phone?: string
  notes?: string
  remark?: string
  confirmedFlag?: number
  confirmed_flag?: number
  participants: GovMeetingParticipant[]
}

export interface GovMeetingAudit {
  id: number
  actionType: string
  action_type?: string
  actionName: string
  action_name?: string
  beforeStatus?: string
  before_status?: string
  afterStatus?: string
  after_status?: string
  opinion?: string
  operatorName?: string
  operator_name?: string
  operatorDeptName?: string
  operator_dept_name?: string
  createdAt?: string
  created_at?: string
}

export interface GovMeetingRecord {
  id: number
  content: string
  conclusions?: string
  followUpItems?: string
  follow_up_items?: string
  recorderName?: string
  recorder_name?: string
  recordTime?: string
  record_time?: string
  createdAt?: string
  created_at?: string
}

export interface GovMeetingEvaluation {
  id: number
  satisfaction: string
  score: number
  resolvedFlag?: number
  resolved_flag?: number
  comment?: string
  createdAt?: string
  created_at?: string
}

export interface GovMeetingDetail extends GovMeetingApply {
  arrangement?: GovMeetingArrangement | null
  auditTrail?: GovMeetingAudit[]
  audit_trail?: GovMeetingAudit[]
  records?: GovMeetingRecord[]
  attachments?: any[]
  evaluation?: GovMeetingEvaluation | null
}

export interface PageResult<T> {
  records: T[]
  total: number
  pageNo: number
  pageSize: number
}

// ── Compat getters ─────────────────────────────────────────────────────────

export function applyNo(a: GovMeetingApply) { return s(a, 'applyNo', 'apply_no') || '' }
export function expectedLevelName(a: GovMeetingApply) { return s(a, 'expectedLevelName', 'expected_level_name') || s(a, 'meetingLevel', 'meeting_level') || '' }
export function finalLevelName(a: GovMeetingApply) { return s(a, 'finalLevelName', 'final_level_name') || '' }
export function applyTitle(a: GovMeetingApply) { return s(a, 'title') || s(a, 'topicName', 'topic_name') || '' }
export function applyContent(a: GovMeetingApply) { return s(a, 'content', 'meeting_content') || s(a, 'description') || '' }
export function applyDiscussion(a: GovMeetingApply) { return s(a, 'discussionItem', 'discussion_item') || '' }
export function urgencyLevel(a: GovMeetingApply) { return s(a, 'urgencyLevel', 'urgency_level') || '' }
export function submittedAt(a: GovMeetingApply) { return s(a, 'submittedAt', 'submitted_at') || '' }
export function acceptedAt(a: GovMeetingApply) { return s(a, 'acceptedAt', 'accepted_at') || '' }

export function arrMeetingDate(arr: GovMeetingArrangement) { return s(arr, 'meetingDate', 'meeting_date') || '' }
export function arrStartTime(arr: GovMeetingArrangement) { return s(arr, 'startTime', 'start_time') || '' }
export function arrEndTime(arr: GovMeetingArrangement) { return s(arr, 'endTime', 'end_time') || '' }
export function arrPlace(arr: GovMeetingArrangement) { return s(arr, 'meetingPlace', 'meeting_place') || '' }
export function arrMethod(arr: GovMeetingArrangement) { return s(arr, 'meetingMethod', 'meeting_method') || '' }
export function arrHostDept(arr: GovMeetingArrangement) { return s(arr, 'hostDeptName', 'host_dept_name') || '' }
export function arrNotes(arr: GovMeetingArrangement) { return s(arr, 'notes') || s(arr, 'remark') || '' }
export function arrParticipants(arr: GovMeetingArrangement): GovMeetingParticipant[] { return arr.participants || [] }

export function pType(p: GovMeetingParticipant) { return s(p, 'participantType', 'participant_type') || '' }
export function pName(p: GovMeetingParticipant) { return s(p, 'participantName', 'participant_name') || '' }
export function pTitle(p: GovMeetingParticipant) { return s(p, 'participantTitle', 'participant_title') || '' }
export function pDept(p: GovMeetingParticipant) { return s(p, 'participantDeptName', 'participant_dept_name') || '' }
export function pPhone(p: GovMeetingParticipant) { return s(p, 'contactPhone', 'contact_phone') || '' }
export function pRole(p: GovMeetingParticipant) { return s(p, 'roleName', 'role_name') || '' }

export function auditActionName(a: GovMeetingAudit) { return s(a, 'actionName', 'action_name') || '' }
export function auditOpinion(a: GovMeetingAudit) { return s(a, 'opinion') || '' }
export function auditOperator(a: GovMeetingAudit) { return s(a, 'operatorName', 'operator_name') || '' }
export function auditCreatedAt(a: GovMeetingAudit) { return s(a, 'createdAt', 'created_at') || '' }
export function detailAuditTrail(d: GovMeetingDetail): GovMeetingAudit[] { return d.auditTrail || d.audit_trail || [] }
export function detailRecords(d: GovMeetingDetail): GovMeetingRecord[] { return d.records || [] }

// ── API functions ─────────────────────────────────────────────────────────

/** 政企约见须知（后端暂无此接口，前端使用兜底文本） */
export async function getGovMeetingNotice(): Promise<string | null> {
  try {
    const res = await get<{ content?: string; html?: string }>('/api/enterprise/gov-meetings/notice')
    return res?.content || res?.html || null
  } catch {
    return null
  }
}

export function createGovMeeting(data: {
  contactName: string
  contactPhone: string
  topicCode?: string
  topicName?: string
  expectedLevelCode?: string
  expectedLevelName?: string
  title?: string
  content?: string
  discussionItem?: string
  urgencyLevel?: string
  industryCode?: string
  industryName?: string
  registeredAddress?: string
  regionCode: string
  regionName: string
  serviceCenterId?: number
  serviceCenterName?: string
  commitmentChecked: number
  attachmentIds?: number[]
}): Promise<GovMeetingApply> {
  return post<GovMeetingApply>('/api/enterprise/gov-meetings', data)
}

export function getMyGovMeetings(params: {
  status?: string
  pageNo?: number
  pageSize?: number
}): Promise<PageResult<GovMeetingApply>> {
  return get<PageResult<GovMeetingApply>>('/api/enterprise/gov-meetings', params as Record<string, unknown>)
}

export function getGovMeetingDetail(id: number): Promise<GovMeetingDetail> {
  return get<GovMeetingDetail>(`/api/enterprise/gov-meetings/${id}`)
}

export function supplementGovMeeting(id: number, data: {
  content?: string
  attachmentIds?: number[]
}): Promise<GovMeetingApply> {
  return post<GovMeetingApply>(`/api/enterprise/gov-meetings/${id}/supplement`, data)
}

/**
 * 企业评价政企约见
 * 后端实际路径：POST /api/enterprise/gov-meetings/{id}/evaluate（非 /evaluation）
 */
export function evaluateGovMeeting(id: number, data: {
  satisfaction: string
  score: number
  resolvedFlag?: number
  comment?: string
}): Promise<GovMeetingApply> {
  return post<GovMeetingApply>(`/api/enterprise/gov-meetings/${id}/evaluate`, data)
}
